### ISSUE TITLE
HttpException - status code field

### ISSUE THREAD

**Issue by matanshukry** [dart-lang/sdk#37423]
This tracker is for issues related to:
* Dart core libraries ("dart:http")

* Dart SDK Version (`dart --version`)
`Dart VM version: 2.4.0 (Wed Jun 19 11:53:45 2019 +0200) on "windows_x64"`

* Whether you are using Windows, MacOSX, or Linux (if applicable)
Windows

* Whether you are using Chrome, Safari, Firefox, Edge (if applicable)
Chrome, although not relevant

HttpException currently only has message and uri fields. Status code fields is a common use case.
My personal use case: https://github.com/flutter/flutter/pull/35335

p.s.
I did notice there's another HttpException in "tools/gardening" (what is that?) ( https://github.com/dart-lang/sdk/blob/2.4.0/tools/gardening/lib/src/util.dart ), with a uri and a status code. Is there a reason there isn't a status code in the sdk exception class?

**Comment by dnfield** [comment]
I'm going to guess this was just an oversight - @zanderso or @bkonyi may know more.  This is probably a pretty trivial thing to fix.

**Comment by bkonyi** [comment]
Yeah, this should be trivial to add but I'd want to chat with @zanderso first to figure out what the status of the http libraries are at this point.

**Comment by matanshukry** [comment]
@zanderso Thoughts?

**Comment by zanderso** [comment]
Internally, the http implementation would never throw an HttpException with a status code since an error status code is already on the HttpResponse etc. and doesn't necessarily indicate an exceptional condition---that would be up to the user of the API to determine. I would recommend continuing with https://github.com/flutter/flutter/pull/35335.

**Comment by dnfield** [comment]
I can see some value in abstracting away the HTTP code for someone who doesn't get access to the response - at the same time, I wonder if doing so would limit a reasonable use case.  Imagine the following:

```dart
try {
var image = await myImageService.get('http://google.com/image.png');
} on SomeException ....
```

What should that SomeException be?  Should ever implementer of such a service have to create a custom exception for it?  

Could the framework just add the StatusCode and it would be null sometimes where it's not used?

**Comment by dnfield** [comment]
Talked offline about this a bit more - HttpException doesn't sound like the right thing for this, but we should probably update the docs on it to clarify that it's for lower level exceptions in the http stack.

Accoridng to @zanderso, it's used when something has gone wrong at a lower level where we don't even have access to a status code.  We'll have to just go ahead and implement this upstream.

**Comment by bkonyi** [comment]
cc @brianquinlan

**Comment by brianquinlan** [comment]
> Talked offline about this a bit more - HttpException doesn't sound like the right thing for this, but we should probably update the docs on it to clarify that it's for lower level exceptions in the http stack.
> 
> Accoridng to @zanderso, it's used when something has gone wrong at a lower level where we don't even have access to a status code. We'll have to just go ahead and implement this upstream.

Is there something actionable. to do here on the dart side?

**Comment by brianquinlan** [comment]
> We're using: https://pub.dev/packages/flutter_cache_manager which seem to be using `http` package internally.
> 
> In a below exception handler we do the string comparison in order to find out 404, which works, but feels like it's not the right way.
> 
> If HttpException had `statusCode` field, that would be great.
> 
> ```dart
>   try {
>     final file = await DefaultCacheManager().getSingleFile(url);
>     return file;
>   } on HttpException catch (e) {
>     final is404 = e.message == "Invalid statusCode: 404";    
>   }
> ```

I don't think that `DefaultCacheManager` should be leaking `HttpException`.
